src/Compilers/CSharp/Portable/Parser/LanguageParser.cs

```diff
@@ -10936,14 +10943,18 @@ bool isParenVarCommaSyntax()
                     if (!this.IsInQuery || !IsTokenQueryContextualKeyword(next))
                     {
                         // Variable must be directly followed by a comma if not followed by exclamation
-                        var afterKind = this.PeekToken(2).Kind;
+                        var after = this.PeekToken(2);
                         // ( x , [...]
-                        if (afterKind == SyntaxKind.CommaToken)
+                        if (after.Kind == SyntaxKind.CommaToken)
                         {
                             return true;
                         }
-                        // ( x! , [...]
-                        if (afterKind == SyntaxKind.ExclamationToken && this.PeekToken(3).Kind == SyntaxKind.CommaToken)
+                        // ( x!! , [...]
+                        if (after.Kind == SyntaxKind.ExclamationToken
```

Comment:
Consider factoring this check out into a local function, and then reusing it above as well.

Reply:
I don't think `!!,` has been used quite enough to warrant a local function, though a oneliner local to check if two tokens are neighbors `isTouching` or `isMergable`, might not be a bad idea.

Reply:
You're doing a bunch more work here to determine whether this is a valid thing to treat as a ParentVarComma, but not doing a similar amount of work above. I'd expect to do the same amount of work in both cases.

---
In reply to: [467122136](https://github.com/dotnet/roslyn/pull/46520#discussion_r467122136) [](ancestors = 467122136)

Reply:
Also, make sure there's a test demonstrating this.

---
In reply to: [467269964](https://github.com/dotnet/roslyn/pull/46520#discussion_r467269964) [](ancestors = 467269964,467122136)

Reply:
Yup, this test (`TestLambdaWithNullValidationParams`) does hit it.